include/revng/Support/GraphAlgorithms.h

```diff
@@ -196,3 +201,215 @@ auto exitless_scc_range(NodeTy Entry) {
 
   return make_filter_range(Range, Filter);
 }
+
+namespace detail {
+
+template<typename T>
+T stripInverseHelper(llvm::Inverse<T> &&);
+
+template<typename T>
+T stripInverseHelper(T &&);
+
+template<typename T>
+using StripInverse = decltype(stripInverseHelper(std::declval<T>()));
+
+// clang-format off
+template<typename T>
+concept SupportedGraph = std::is_pointer_v<T> ||
+                         (StrictSpecializationOf<T, llvm::Inverse> &&
+                          std::is_pointer_v<StripInverse<T>>);
+// clang-format on
+
+template<typename GraphType>
+using GT = llvm::GraphTraits<GraphType>;
+
+} // namespace detail
+
+template<detail::SupportedGraph RawGraphType>
+class GraphNodeOrRootHelper {
+public:
+  using GraphType = detail::StripInverse<RawGraphType>;
+
+  using NodeRef = typename detail::GT<RawGraphType>::NodeRef;
+  using ValueType = std::variant<GraphType, NodeRef>;
+  using EquivalentPair = std::pair<GraphType, NodeRef>;
+
+  using GraphTrait = detail::GT<RawGraphType>;
+  using NodeTrait = std::conditional_t<std::is_same_v<GraphType, RawGraphType>,
+                                       detail::GT<NodeRef>,
+                                       detail::GT<llvm::Inverse<NodeRef>>>;
+
+  using Iterator = VariedIterator<typename GraphTrait::nodes_iterator,
+                                  typename NodeTrait::ChildIteratorType>;
+
+public:
+  ValueType Value;
+
+public:
+  GraphNodeOrRootHelper(GraphType Graph) : Value(Graph) {}
+  GraphNodeOrRootHelper(llvm::Inverse<GraphType> Inv) : Value(Inv.Graph) {}
+  GraphNodeOrRootHelper(NodeRef Node) : Value(std::move(Node)) {}
+  GraphNodeOrRootHelper(std::nullptr_t) : Value((GraphType) nullptr) {}
+  GraphNodeOrRootHelper(EquivalentPair P) : Value((GraphType) nullptr) {
+    if (P.first != nullptr)
+      Value = P.first;
+    else
+      Value = P.second;
+  }
+
+public:
+  Iterator child_begin() {
+    auto Visitor = OverloadedLambda{
+      [](GraphType Graph) { return Iterator(GraphTrait::nodes_begin(Graph)); },
+      [](NodeRef Node) { return Iterator(NodeTrait::child_begin(Node)); }
+    };
+    return std::visit(Visitor, Value);
+  }
+  Iterator child_end() {
+    auto Visitor = OverloadedLambda{
+      [](GraphType Graph) { return Iterator(GraphTrait::nodes_end(Graph)); },
+      [](NodeRef Node) { return Iterator(NodeTrait::child_end(Node)); }
+    };
+    return std::visit(Visitor, Value);
+  }
```

Comment:
You should take `NodeTrait` as a template argument and default it to `GraphTraits<RawGraphType>`.
This way, a user can pass you `GraphTraits<Inverse<RawGraphType>>` or a filtered graph view or whatever else.

Reply:
But then we have a problem of trying to convey to the user that we have two distinct trait requirements - one for the graph and one for the node. I haven't seen any llvm algorithm do that (since most of them just accept a node). We need to be careful since there's enclosing such an argument would likely mean improper values passed in, and in this case, it's very likely to lead to super-hard-to-process error messages.

Reply:
I understand, but then you won't be compatible with `FilteredGraphTraits` and friends.
Frankly, the only way not to get obscure errors, is for the original author of this stuff to provide working examples for the most common use cases.

Anyway, no biggie. Up to you, let's try to move forward.

Reply:
Yeah, ok, then let's hold off on this decision for a bit. After @pfez gets back, maybe he knows something I missed about how we can make this work in a nicer way.

Reply:
Closing this, since I found a way to get rid of this without losing functionality.